Getting started guides for Kyma (#9296)
kyma-project/kyma: docs/getting-started/12-bind-redis-service-to-function.md

---
title: Bind a Redis ServiceInstance to the Function
type: Getting Started
---

This guide shows how you can bind a sample instance of the Redis service to the Function. After completing all steps, you will get the Function with encoded Secrets to the service. You will use Redis as an external database for the Function that replaces its in-memory storage.

## Steps

### Bind the Redis ServiceInstance to the Function

Follows these steps:

<div tabs name="steps" group="bind-redis-to-function">
  <details>
  <summary label="kubectl">
  kubectl
  </summary>

> 1. Create a ServiceBinding CR that points to the existing Redis instance in the **spec.instanceRef** field:

Author: I changed it into:
`Create a ServiceBinding custom resource (CR). Its **spec.instanceRef** field points to your Redis instance:`

> 1. Create a ServiceBinding CR that points to the existing Redis instance in the **spec.instanceRef** field:

Reviewer: This thread is continued [here](https://github.com/kyma-project/kyma/pull/9296/files#r483584261).

> 1. Create a ServiceBinding custom resource (CR). Its **spec.instanceRef** field points to your Redis instance:

Reviewer: Continuing [this](https://github.com/kyma-project/kyma/pull/9296/#discussion_r481983803) thread.

I believe this should stress more that you need to specify the Redis instance in this field.

If you don't want to go with my previous suggestion, 'd either leave your original version just unfolding the acronym: 
```suggestion
1. Create a ServiceBinding custom resource (CR) that points to the existing Redis instance in the **spec.instanceRef** field:
```

or change it to something like this: 
```
1. Create a ServiceBinding custom resource (CR) with its **spec.instanceRef** field pointing to your Redis instance:
```

  ```yaml
  cat <<EOF | kubectl apply -f -
  apiVersion: servicecatalog.k8s.io/v1beta1
  kind: ServiceBinding
  metadata:
    name: orders-function
    namespace: orders-service
  spec:
    instanceRef:
      name: redis-service
  EOF
  ```

2. Check that the ServiceBinding CR was created. The last condition in the CR status should be `Ready True`:

  ```bash
  kubectl get servicebinding orders-function -n orders-service -o=jsonpath="{range .status.conditions[*]}{.type}{'\t'}{.status}{'\n'}{end}"
  ```

3. Create a ServiceBindingUsage CR:

  ```yaml
  cat <<EOF | kubectl apply -f -
  apiVersion: servicecatalog.kyma-project.io/v1alpha1
  kind: ServiceBindingUsage
  metadata:
    name: orders-function
    namespace: orders-service
  spec:
    serviceBindingRef:
      name: orders-function
    usedBy:
      kind: serverless-function
      name: orders-function
    parameters:
      envPrefix:
        name: "REDIS_"
  EOF
  ```

>    - The **spec.serviceBindingRef** and **spec.usedBy** fields are required. **spec.serviceBindingRef** points to the ServiceBinding you have just created and **spec.usedBy** points to the Function. More specifically, **spec.usedBy** refers to the name of the Function and the cluster-specific [UsageKind CR](/components/service-catalog/#custom-resource-usage-kind) (`kind: serverless-function`) that defines how Secrets should be injected to your Function when creating a ServiceBinding.

Author: I would leave it as it is since the items in the 1st bullet point are related and a kind of explanation to the field as first.

>    - The **spec.serviceBindingRef** and **spec.usedBy** fields are required. **spec.serviceBindingRef** points to the ServiceBinding you have just created and **spec.usedBy** points to the Function. More specifically, **spec.usedBy** refers to the name of the Function and the cluster-specific [UsageKind CR](/components/service-catalog/#custom-resource-usage-kind) (`kind: serverless-function`) that defines how Secrets should be injected to your Function when creating a ServiceBinding.

Reviewer: They would still be a part of this explanation and grouped under the first bullet point as related, but as I said, it's just a thought :)

   - The **spec.parameters.envPrefix.name** field is optional. It adds a prefix to all environment variables injected in a Secret to the Function when creating a ServiceBinding. In our example, **envPrefix** is `REDIS_`, so all environmental variables will follow the `REDIS_{env}` naming pattern.

     > **TIP:** It is considered good practice to use **envPrefix**. In some cases, a Function must use several instances of a given ServiceClass. Prefixes allow you to distinguish between instances and make sure that one Secret does not overwrite another one.

> 4. Check that the ServiceBindingUsage CR was created. The last condition in the CR status should be `Ready True`:

Reviewer: Oops... I just spotted the typo (`tis`). Sorry for that. Added a correction [here](https://github.com/kyma-project/kyma/pull/9296/files#r483588129).

> 4. Check that the ServiceBindingUsage CR was created. Tis is indicated by the last condition in the CR status equal to `Ready True`:

Reviewer: ```suggestion
4. Check that the ServiceBindingUsage CR was created. This is indicated by the last condition in the CR status equal to `Ready True`:
```
My bad...
![image](https://user-images.githubusercontent.com/53601578/92239869-faaa9600-eebb-11ea-8f21-02b81ef865dd.png)

  PORT: 6379
  REDIS_PASSWORD: 1tvDcINZvp
  ```

  </details>
  <details>
  <summary label="console-ui">
  Console UI
  </summary>

1. Go to **Development** > **Functions** in the left navigation panel and select `orders-function`.

2. Switch to the **Configuration** tab and select **Create Service Binding** in the **Service Bindings** section.

3. Select `redis-service` from the **Service Instance** drop-down list, add `REDIS_` as **Prefix for injected variables**, and make sure **Create new Secret** is selected.

   > **NOTE:** The **Prefix for injected variables** field is optional. On creating a ServiceBinding, it adds a prefix to all environment variables injected in a Secret to the Function. In our example, the prefix is set to `REDIS_`, so all environment variables will follow the `REDIS_{ENVIRONMENT_VARIABLE}` naming pattern.

   > **TIP:** It is considered good practice to use prefixes for environment variables. In some cases, a Function must use several instances of a given ServiceClass. Prefixes allow you to distinguish between instances and make sure that one Secret does not overwrite another.

4. Select **Create** to confirm the changes.

> A message confirming that the ServiceBinding was created will appear in the **Service Bindings** section in your Function along with environment variable names.

Reviewer: ```suggestion
A message confirming that the ServiceBinding was created will appear in the **Service Bindings** section in your Function's view along with environment variable names.
```
I just thought, should we add `view` here? We don't really see it in the Function itself, right?

Author: You're right :)

If you switch to the **Code** tab and scroll down to the **Environment Variables** section, you should see `REDIS_PORT`, `REDIS_HOST` and `REDIS_REDIS_PASSWORD` items with the `Service Binding` type. It indicates that the environment variable was injected to the Function by the ServiceBinding.

    </details>
</div>

### Call and test the Function

Follow these steps:

> > **CAUTION:** If you have a Minikube cluster, you must first add the IP address of the exposed Service to the `hosts` file on your machine:

Reviewer: ```suggestion
> **CAUTION:** If you have a Minikube cluster, you must first add its IP address to the `hosts` file on your machine:
```

or ```(...) add its IP address mapped to the domain of the exposed Service to the `hosts` file on your machine:```

What's really added to `/etc/hosts` is the Minikube cluster's IP (`minikube ip`) mapped to the service's hostname/domain (`orders-function.kyma.local`) so that accessing the domain can redirect you to the cluster.

>
>  ```bash
>  echo "$(minikube ip) orders-service.kyma.local" | sudo tee -a /etc/hosts
>  ```

1. Retrieve the domain of the exposed Function and save it to the environment variable:

   ```bash
   export FUNCTION_DOMAIN=$(kubectl get virtualservices -l apirule.gateway.kyma-project.io/v1alpha1=orders-function.orders-service -n orders-service -o=jsonpath='{.items[*].spec.hosts[0]}')
   ```

2. Run this command in the terminal to call the Function:

   ```bash
   curl -ik "https://$FUNCTION_DOMAIN"
   ```

   The system returns a response similar to this one:

   ```bash
   HTTP/2 200
   access-control-allow-origin: *
   content-length: 187
   content-type: application/json; charset=utf-8
   date: Mon, 13 Jul 2020 13:05:33 GMT
   etag: W/"bb-Iuoc/aX9UjdqADJAZPNrwPdoraI"
   server: istio-envoy
   x-envoy-upstream-service-time: 25
   x-powered-by: Express

   [{"orderCode":"762727210","consignmentCode":"76272725","consignmentStatus":"PICKUP_COMPLETE"}, {"orderCode":"123456789","consignmentCode":"76272725","consignmentStatus":"PICKUP_COMPLETE"}]
   ```

   > **NOTE:** The listed orders are those created in the previous guides.

> 3. Send a `POST` request to the Function with a sample order details:

Reviewer: ```suggestion
3. Send a `POST` request to the Function with sample order details:
```
See [this comment](https://github.com/kyma-project/kyma/pull/9296#discussion_r480108879).

   ```bash
   curl -ikX POST "https://$FUNCTION_DOMAIN" \
     -H "Content-Type: application/json" \
     -H 'Cache-Control: no-cache' -d \
     '{
       "orderCode": "762727234",
       "consignmentCode": "76272725",
       "consignmentStatus": "PICKUP_COMPLETE"
     }'
   ```

4. When you call the `https://$FUNCTION_DOMAIN` URL again, the system returns a response similar to this one:

   ```bash
   HTTP/2 200
   access-control-allow-origin: *
   content-length: 280
   content-type: application/json; charset=utf-8
   date: Mon, 13 Jul 2020 13:05:51 GMT
   etag: W/"118-lAc/HJqUaWFKlQ/uyvrhuPb++80"
   server: istio-envoy
   x-envoy-upstream-service-time: 9
   x-powered-by: Express

   [{"orderCode":"762727234","consignmentCode":"76272725","consignmentStatus":"PICKUP_COMPLETE"}, {"orderCode":"762727210","consignmentCode":"76272725","consignmentStatus":"PICKUP_COMPLETE"}, {"orderCode":"123456789","consignmentCode":"76272725","consignmentStatus":"PICKUP_COMPLETE"}]
   ```

   You can see the Function returns the previously sent order details.

> 5. Remove the [Pod](https://kubernetes.io/docs/concepts/workloads/pods/) created by the `orders-function` Function. Run this command and wait for the system to delete the Pod and start a new one:

Reviewer: ```suggestion
5. Remove the [Pod](https://kubernetes.io/docs/concepts/workloads/pods/) created by the `orders-function` Function. Run this command, and wait for the system to delete the Pod and start a new one:
```
Again, I added the comma to separate the two `and`'s.

   ```bash
   kubectl delete pod -n orders-service -l "serverless.kyma-project.io/function-name=orders-function"
   ```

6. Call the Function again to check the storage:

   ```bash
   curl -ik "https://$FUNCTION_DOMAIN"
   ```

   The system returns a response similar to this one:

   ```bash
   HTTP/2 200
   access-control-allow-origin: *
   content-length: 280
   content-type: application/json; charset=utf-8
   date: Mon, 13 Jul 2020 13:05:33 GMT
   etag: W/"118-lAc/HJqUaWFKlQ/uyvrhuPb++80"
   server: istio-envoy
   x-envoy-upstream-service-time: 975
   x-powered-by: Express

   [{"orderCode":"762727234","consignmentCode":"76272725","consignmentStatus":"PICKUP_COMPLETE"}, {"orderCode":"762727210","consignmentCode":"76272725","consignmentStatus":"PICKUP_COMPLETE"}, {"orderCode":"123456789","consignmentCode":"76272725","consignmentStatus":"PICKUP_COMPLETE"}]
   ```

>    As you can see, `orders-function` uses Redis storage now. It means that every time you delete the Pod of the Function or change its Deployment definition, the order details will not get lost.

Reviewer: ```suggestion
   As you can see, `orders-function` uses Redis storage now. It means that every time you delete the Pod of the Function or change its Deployment definition, the order details are preserved.
```
See [this comment](https://github.com/kyma-project/kyma/pull/9296/files#r480121553) (not the suggestion, just how you phrased the line), and [this comment](https://github.com/kyma-project/kyma/pull/9296#discussion_r478412987) for the reason why I changed it.

Follow these steps:

> > **CAUTION:** If you have a Minikube cluster, you must first add its IP address mapped to the exposed Kubernetes Service to the `hosts` file on your machine:

Reviewer: ```suggestion
> **CAUTION:** If you have a Minikube cluster, you must first add its IP address mapped to the hostname of the exposed Kubernetes Service to the `hosts` file on your machine:
```
